src/Razor/src/Microsoft.AspNetCore.Razor.VSCode/src/RazorRenameFeature.ts

```diff
@@ -0,0 +1,97 @@
+/* --------------------------------------------------------------------------------------------
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See License.txt in the project root for license information.
+ * ------------------------------------------------------------------------------------------ */
+
+import * as vscode from 'vscode';
+import { getRazorDocumentUri, isRazorCSharpFile } from './RazorConventions';
+import { RazorDocumentManager } from './RazorDocumentManager';
+import { RazorDocumentSynchronizer } from './RazorDocumentSynchronizer';
+import { RazorLanguage } from './RazorLanguage';
+import { RazorLanguageServiceClient } from './RazorLanguageServiceClient';
+import { RazorLogger } from './RazorLogger';
+import { RazorRenameProvider } from './RazorRenameProvider';
+import { LanguageKind } from './RPC/LanguageKind';
+
+export class RazorRenameFeature {
+    private readonly renameProvider: RazorRenameProvider;
+
+    constructor(
+        documentSynchronizer: RazorDocumentSynchronizer,
+        documentManager: RazorDocumentManager,
+        private readonly serviceClient: RazorLanguageServiceClient,
+        logger: RazorLogger) {
+            this.renameProvider = new RazorRenameProvider(
+                documentSynchronizer,
+                documentManager,
+                serviceClient,
+                logger);
+    }
+
+    public register() {
+        const registrations = [
+            vscode.languages.registerRenameProvider(RazorLanguage.id, this.renameProvider),
+
+            // Register a command that will be called whenever an edit happens to any CSharp file.
+            // It takes in the edits that Omnisharp is planning to do and re-map any Razor edits to the
+            // original Razor documents.
+            vscode.commands.registerCommand(
+                'razor.remapRenameEdit', async (
+                    document: vscode.TextDocument,
+                    position: vscode.Position,
+                    newText: string,
+                    cSharpEdit: vscode.WorkspaceEdit) => {
+                        const remappedEdit = await this.remapRazorCSharpEdits(
+                        cSharpEdit,
+                        this.serviceClient);
+                        return remappedEdit;
+                    }),
+        ];
+
+        return vscode.Disposable.from(...registrations);
+    }
+
+    // Given a set of C# edits, we need to map edits to Razor CSharp documents back to the original Razor documents.
+    private async remapRazorCSharpEdits(
```

Comment:
Should rename this to `remapWorkspaceEdit`, looking at the spec workspace edits are used all over the place so it'll be good for this to be available for all workspace edits.

We'll probably eventually want to break this piece of code into its own `remapService`. That's for another day though.

Reply:
Sure I can do that.